File: android_native_reference/FaceLandmarksPlugin.kt
// Reference VisionCamera Frame Processor Plugin: detects a face in one
// camera frame and returns a flat record of key landmark points (eye
// centers/lids, mouth corners) for src/signal/faceSymmetryAnalysis.ts to
// compare left vs right. Uses Google ML Kit Face Detection (on-device,
// bundled model — no Firebase project or network call required).
//
// This is a starting point, not a drop-in file — wire it up per VisionCamera's
// "Creating Frame Processor Plugins" guide, same as MeanRedChannelPlugin.kt.
// Also add to android/app/build.gradle:
//   implementation("com.google.mlkit:face-detection:16.1.7")
//
// package: com.veritahealth.frameprocessors (adjust to your app id)

package com.veritahealth.frameprocessors

import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceContour
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.face.FaceLandmark
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy
import java.util.concurrent.TimeUnit
import kotlin.math.max
import kotlin.math.min

class FaceLandmarksPlugin(proxy: VisionCameraProxy, options: Map<String, Any>?) :
    FrameProcessorPlugin() {

    private val detectorOptions = FaceDetectorOptions.Builder()
        .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
        .setContourMode(FaceDetectorOptions.CONTOUR_MODE_ALL)
        .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
        .build()

    private val detector = FaceDetection.getClient(detectorOptions)

    override fun callback(frame: Frame, arguments: Map<String, Any>?): Any {
        val notDetected = mapOf("faceDetected" to false)
        val rotationDegrees = frame.orientation.toDegrees()
        val inputImage = InputImage.fromMediaImage(frame.image, rotationDegrees)

        val faces: List<Face> = try {
            // ML Kit's detector is Task-based (async); frame processors run on
            // their own background thread already, so blocking here briefly
            // is the standard pattern for synchronous ML Kit frame processor
            // plugins. A short timeout avoids stalling the pipeline if the
            // model hiccups on one frame.
            Tasks.await(detector.process(inputImage), 1000, TimeUnit.MILLISECONDS)
        } catch (e: Exception) {
            return notDetected
        }

        val face = faces.firstOrNull() ?: return notDetected

        val leftEyeContour = face.getContour(FaceContour.LEFT_EYE)?.points
        val rightEyeContour = face.getContour(FaceContour.RIGHT_EYE)?.points
        val leftMouth = face.getLandmark(FaceLandmark.MOUTH_LEFT)?.position
        val rightMouth = face.getLandmark(FaceLandmark.MOUTH_RIGHT)?.position
        val leftEyeLandmark = face.getLandmark(FaceLandmark.LEFT_EYE)?.position
        val rightEyeLandmark = face.getLandmark(FaceLandmark.RIGHT_EYE)?.position

        if (leftEyeContour.isNullOrEmpty() || rightEyeContour.isNullOrEmpty() ||
            leftMouth == null || rightMouth == null || leftEyeLandmark == null || rightEyeLandmark == null
        ) {
            return notDetected
        }

        val leftEyeTop = leftEyeContour.maxByOrNull { it.y }!! // smaller y = higher on screen in ML Kit's image space
        val leftEyeBottom = leftEyeContour.minByOrNull { it.y }!!
        val rightEyeTop = rightEyeContour.maxByOrNull { it.y }!!
        val rightEyeBottom = rightEyeContour.minByOrNull { it.y }!!

        return mapOf(
            "faceDetected" to true,
            "leftEyeCenterX" to leftEyeLandmark.x, "leftEyeCenterY" to leftEyeLandmark.y,
            "rightEyeCenterX" to rightEyeLandmark.x, "rightEyeCenterY" to rightEyeLandmark.y,
            "leftEyeTopX" to leftEyeTop.x, "leftEyeTopY" to leftEyeTop.y,
            "leftEyeBottomX" to leftEyeBottom.x, "leftEyeBottomY" to leftEyeBottom.y,
            "rightEyeTopX" to rightEyeTop.x, "rightEyeTopY" to rightEyeTop.y,
            "rightEyeBottomX" to rightEyeBottom.x, "rightEyeBottomY" to rightEyeBottom.y,
            "leftMouthCornerX" to leftMouth.x, "leftMouthCornerY" to leftMouth.y,
            "rightMouthCornerX" to rightMouth.x, "rightMouthCornerY" to rightMouth.y,
        )
    }
}

// Registration (typically in your ReactPackage's createFrameProcessorPlugins,
// or via VisionCameraProxy's plugin registry per the current VisionCamera version):
//
//   VisionCameraProxy.getInstance(reactContext).initFrameProcessorPlugin(
//       "detectFaceLandmarks"
//   ) { proxy, options -> FaceLandmarksPlugin(proxy, options) }

File: ios_native_reference/FaceLandmarksPlugin.swift
// Reference VisionCamera Frame Processor Plugin: detects a face in one
// camera frame and returns a flat record of key landmark points (eye
// centers/lids, mouth corners) for src/signal/faceSymmetryAnalysis.ts to
// compare left vs right. Uses Apple's Vision framework — no extra
// dependency or CocoaPod needed, it ships with iOS.
//
// This is a starting point, not a drop-in file — register it per VisionCamera's
// "Creating Frame Processor Plugins" guide, same as MeanRedChannelPlugin.swift.
//
// Coordinate note: `pointsInImage(imageSize:)` returns points in the full
// frame's coordinate space, which is all this plugin's consumer needs —
// src/signal/faceSymmetryAnalysis.ts only ever compares distances *within*
// one frame, so it doesn't matter whether that space is nominally top-left
// or bottom-left origin as long as it's consistent, which Vision guarantees.
//
// Mirroring note: the front camera's raw buffer is not necessarily mirrored
// the way the on-screen preview is. "leftMouthCorner" here means "the
// corner at the smaller X coordinate in the raw frame", which may or may
// not match the subject's anatomical left depending on orientation/mirroring
// — verify this on a real device before trusting the left/right labels in
// UI copy that names a side. It does not affect whether asymmetry is
// detected, only which side gets named.

import VisionCamera
import Vision
import CoreVideo

@objc(FaceLandmarksPlugin)
public class FaceLandmarksPlugin: FrameProcessorPlugin {

  public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any {
    guard let pixelBuffer = frame.buffer else {
      return ["faceDetected": false]
    }

    let imageSize = CGSize(
      width: CVPixelBufferGetWidth(pixelBuffer),
      height: CVPixelBufferGetHeight(pixelBuffer)
    )

    let request = VNDetectFaceLandmarksRequest()
    let handler = VNImageRequestHandler(cvPixelBuffer: pixelBuffer, orientation: .up, options: [:])

    do {
      try handler.perform([request])
    } catch {
      return ["faceDetected": false]
    }

    guard
      let face = request.results?.first as? VNFaceObservation,
      let landmarks = face.landmarks
    else {
      return ["faceDetected": false]
    }

    // Landmark regions are normalized to the face bounding box; convert to
    // full-frame coordinates so left/right comparisons share one frame of
    // reference (needed since the two eyes/mouth corners come from
    // different, independently-normalized regions).
    func imagePoints(_ region: VNFaceLandmarkRegion2D?) -> [CGPoint] {
      guard let region = region else { return [] }
      let boxOrigin = face.boundingBox.origin
      let boxSize = face.boundingBox.size
      return region.normalizedPoints.map { p in
        CGPoint(
          x: (boxOrigin.x + p.x * boxSize.width) * imageSize.width,
          y: (boxOrigin.y + p.y * boxSize.height) * imageSize.height
        )
      }
    }

    func center(_ points: [CGPoint]) -> CGPoint? {
      guard !points.isEmpty else { return nil }
      let sum = points.reduce(CGPoint.zero) { CGPoint(x: $0.x + $1.x, y: $0.y + $1.y) }
      return CGPoint(x: sum.x / CGFloat(points.count), y: sum.y / CGFloat(points.count))
    }

    let leftEyePoints = imagePoints(landmarks.leftEye)
    let rightEyePoints = imagePoints(landmarks.rightEye)
    let mouthPoints = imagePoints(landmarks.outerLips)

    guard
      let leftEyeCenter = center(leftEyePoints),
      let rightEyeCenter = center(rightEyePoints),
      let leftEyeTop = leftEyePoints.max(by: { $0.y < $1.y }),
      let leftEyeBottom = leftEyePoints.min(by: { $0.y < $1.y }),
      let rightEyeTop = rightEyePoints.max(by: { $0.y < $1.y }),
      let rightEyeBottom = rightEyePoints.min(by: { $0.y < $1.y }),
      let leftMouthCorner = mouthPoints.min(by: { $0.x < $1.x }),
      let rightMouthCorner = mouthPoints.max(by: { $0.x < $1.x })
    else {
      return ["faceDetected": false]
    }

    return [
      "faceDetected": true,
      "leftEyeCenterX": leftEyeCenter.x, "leftEyeCenterY": leftEyeCenter.y,
      "rightEyeCenterX": rightEyeCenter.x, "rightEyeCenterY": rightEyeCenter.y,
      "leftEyeTopX": leftEyeTop.x, "leftEyeTopY": leftEyeTop.y,
      "leftEyeBottomX": leftEyeBottom.x, "leftEyeBottomY": leftEyeBottom.y,
      "rightEyeTopX": rightEyeTop.x, "rightEyeTopY": rightEyeTop.y,
      "rightEyeBottomX": rightEyeBottom.x, "rightEyeBottomY": rightEyeBottom.y,
      "leftMouthCornerX": leftMouthCorner.x, "leftMouthCornerY": leftMouthCorner.y,
      "rightMouthCornerX": rightMouthCorner.x, "rightMouthCornerY": rightMouthCorner.y,
    ]
  }
}

// Registration (typically in AppDelegate or a dedicated setup file):
//
//   FrameProcessorPluginRegistry.addFrameProcessorPlugin("detectFaceLandmarks") { proxy, options in
//     FaceLandmarksPlugin(proxy: proxy, options: options)
//   }
